mlflow/pyfunc/backend.py

```diff
@@ -86,8 +86,12 @@ def __init__(
         super().__init__(config=config, **kwargs)
         self._nworkers = workers or 1
         if env_manager == _EnvManager.CONDA and ENV not in config:
+            warnings.warn(
+                "Conda environment is not specified in config `env`. Using local environment."
+            )
             env_manager = _EnvManager.LOCAL
-        self._env_manager = env_manager
+        self._user_env_manager = env_manager
```

Comment:
Hmm the handling of two instance variables look confusing. It's not clear what values will be set for what use case.

Would it be possible to update the implementation to directly represent what we want to do?
1. If `env_manager` is None
    1.  If python base image, default to `local`
    2. Otherwise, default to `virtualenv`
2. If passed `env_manager` is `local`
    1.  If we can use python base image, use it.
    2. Otherwise, raise exception and prompt users to specify either `conda` or `virtualenv`
3. If passed `env_manager` is `conda` or `virtualenv`
    1. Always use ubuntu base image.

Reply:
I think the challenge of keeping a single param is that we need to update everywhere else using self._env_manager to something like `self._env_manager or virtualenv`

Reply:
I see. I might be wrong, but from quick search, the `env_manager` arg should be always passed from upstream via `get_flavor_backend()` call. Could you double-check if that's the case? Then we can simply make it required argument.

Reply:
Oh I mean `self._env_manager` used in PyFuncBackend methods, for example `predict` and `prepare_env`

Reply:
So the `self._env_manager` is passed to backend from the **upstream** function like [mlflow.models.python_api.predict()](https://github.com/mlflow/mlflow/blob/6501d3844c531bd8e73fc23dfb8a1b8c66ce8db6/mlflow/models/python_api.py#L99), and then referred by those instance method. Since the upstream function defines their own default and always specify `_env_manager`, we don't have to implement a single defaulting logic in `PyFuncBackend`.

Reply:
I see, that makes sense, but from the code perspective it's possible someone misuses the init method by not setting a default _env_manager value. Or do you think it's fine? As it will raise exception if the _env_manager is not defined anyways

Reply:
> it's possible someone misuses the init method

If we remove the default value in constructor, it will be mandatory to set some value in `env_manager`:)